service/header/store.go

```diff
@@ -0,0 +1,302 @@
+const (
+	// DefaultStoreCache defines the amount of max entries allowed in the Header Store cache.
+	DefaultStoreCache = 1024
+	// DefaultIndexCache defines the amount of max entries allowed in the Height to Hash index cache.
+	DefaultIndexCache = 256
+)
+
+var (
+	storePrefix = datastore.NewKey("headers")
+	headKey     = datastore.NewKey("head")
+)
+
+type store struct {
+	ds    datastore.Batching
+	cache *lru.ARCCache
+	index *heightIndexer
+
+	headLk sync.RWMutex
+	head   bytes.HexBytes
+}
+
+func NewStore(ds datastore.Batching) (Store, error) {
+	return newStore(ds)
+}
+
+func NewStoreWithHead(ds datastore.Batching, head *ExtendedHeader) (Store, error) {
+	store, err := newStore(ds)
+	if err != nil {
+		return nil, err
+	}
+
+	err = store.put(head)
+	if err != nil {
+		return nil, err
+	}
+
+	return store, store.newHead(head.Hash())
+}
+
+func newStore(ds datastore.Batching) (*store, error) {
+	ds = namespace.Wrap(ds, storePrefix)
+	cache, err := lru.NewARC(DefaultStoreCache)
+	if err != nil {
+		return nil, err
+	}
+
+	index, err := newHeightIndexer(ds)
+	if err != nil {
+		return nil, err
+	}
+
+	return &store{
+		ds:    ds,
+		cache: cache,
+		index: index,
+	}, nil
+}
+
+func (s *store) Start(context.Context) error {
+	err := s.loadHead()
+	if err == datastore.ErrNotFound {
+		return fmt.Errorf("header/store: no chain head")
+	}
+
+	return err
+}
+
+func (s *store) Head(ctx context.Context) (*ExtendedHeader, error) {
+	s.headLk.RLock()
+	head := s.head
+	s.headLk.RUnlock()
+	return s.Get(ctx, head)
+}
+
+func (s *store) Get(_ context.Context, hash bytes.HexBytes) (*ExtendedHeader, error) {
+	if v, ok := s.cache.Get(hash); ok {
+		return v.(*ExtendedHeader), nil
+	}
+
+	b, err := s.ds.Get(datastore.NewKey(hash.String()))
+	if err != nil {
+		if err == datastore.ErrNotFound {
+			return nil, ErrNotFound
+		}
+
+		return nil, err
+	}
+
+	return UnmarshalExtendedHeader(b)
+}
+
+func (s *store) GetByHeight(ctx context.Context, height int64) (*ExtendedHeader, error) {
+	hash, err := s.index.HashByHeight(height)
+	if err != nil {
+		if err == datastore.ErrNotFound {
+			return nil, ErrNotFound
+		}
+
+		return nil, err
+	}
+
+	return s.Get(ctx, hash)
+}
+
+func (s *store) GetRangeByHeight(ctx context.Context, from, to int64) ([]*ExtendedHeader, error) {
+	h, err := s.GetByHeight(ctx, to)
+	if err != nil {
+		return nil, err
+	}
+
+	ln := to - from
+	headers := make([]*ExtendedHeader, ln)
+	for i := ln - 1; i > 0; i-- {
+		headers[i] = h
+		h, err = s.Get(ctx, h.LastHeader())
+		if err != nil {
+			return nil, err
+		}
+	}
+	headers[0] = h
+
+	return headers, nil
+}
+
+func (s *store) Has(_ context.Context, hash bytes.HexBytes) (bool, error) {
+	if ok := s.cache.Contains(hash); ok {
+		return ok, nil
+	}
+
+	return s.ds.Has(datastore.NewKey(hash.String()))
+}
+
+func (s *store) Append(ctx context.Context, headers ...*ExtendedHeader) error {
```

Comment:
I think this deserves a comment that the passed-in headers are expected to be adjacent / a chain of consecutive headers.

Reply:
True

Reply:
Will add to the comment to the interface

Reply:
Done